### PR TITLE
differentiate context for start and wait

### PR DESCRIPTION
**Pull request by haitch**
- context that pass to Start() will determin task lifetime
- context that pass to Wait() will not impact task lifetime

### CODE REVIEW

File: async_task.go [hunk 1]
```diff
@@ -88,12 +90,12 @@ func (t *TaskStatus) Wait(ctx context.Context) (interface{}, error) {
...
 	case <-ctx.Done():
-		t.finish(StateCanceled, nil, ErrTimeout)
-		return t.result, t.err
+		return nil, ErrWaitTimeout
 	}
 }
 
 // WaitWithTimeout block current thread/routine until task finished or failed, or exceed the duration specified.
+// timeout only stop waiting, taks will remain running.
 func (t *TaskStatus) WaitWithTimeout(ctx context.Context, timeout time.Duration) (interface{}, error) {
```
Comment (serbrech) [inline async_task.go:99]: why wouldn't you rely on the context to setup the timeout? 
`context.WithTimeout(context.Background(), 20*time.Second)`

Comment (serbrech) [reply async_task.go:99]: **suggestion** : leave it to the caller to setup a timeout on the context passed in.

File: async_task.go [hunk 2]
```diff
@@ -88,12 +90,12 @@ func (t *TaskStatus) Wait(ctx context.Context) (interface{}, error) {
...
 	// return immediately if task already in terminal state.
 	if t.state.IsTerminalState() {
```
Comment (serbrech) [inline async_task.go:101]: **question**: this is already implemented in Wait(...) do you need it here?

Comment (serbrech) [reply async_task.go:101]: I can't comment on the below lines, but, you are calling `defer cancelFunc()` that's also called in the Wait() implementation, is shouldn't that all be just handled in Wait()?

Comment (serbrech) [reply async_task.go:101]: otherwise the cancelfunc is called twice? maybe write a test to check that assumption?

Comment (haitch) [reply async_task.go:101]: this defer cancelFunc() is the cancelFunc from WithTimeout.

in Wait the cancelFunc is the context started the task

File: async_task.go (hunk 1 above)
Comment (haitch) [reply async_task.go:99]: the caller can definitely call Wait() with a context already setup.

or caller can invoke this shortcut, which make code easy to read.

```
ctxForWait, cancelFunc := context.WithTimeout(ctx, 2*time.Second)
defer cancelFunc() // or you get a warning for ignore this cancelFunc
tsk.Wait(ctxForWait)
```
vs 
```
context.WaitWithTimeout(ctx, 2*time.Second)
```

File: error_test.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ func TestTimeoutCase(t *testing.T) {
 	ctx := newTestContext(t)
 	tsk := asynctask.Start(ctx, getCountingTask(200*time.Millisecond))
 	_, err := tsk.WaitWithTimeout(ctx, 300*time.Millisecond)
-	assert.True(t, errors.Is(err, asynctask.ErrWaitTimeout), "expecting ErrWaitTimeout")
+	assert.True(t, errors.Is(err, context.DeadlineExceeded), "expecting DeadlineExceeded")
```
Comment (serbrech) [inline error_test.go:44]: 👍

**Pull request merged**